SS01 - 장소 검색 기능 (feature/ss01-Retrofit)

nusurprise opened the pull request:
## 개요
- 장소 검색 기능(검색 요청 기능, 검색 결과 출력 기능, 검색 장소 선택 기능, 뒤로 가기 기능) 화면 구현
## 작업 내용
- 장소 검색을 위한 LocationSearchActivity, Presenter, Contract, LocationSearchAdapter 구현 
- Retrofit 호출 처리를 위한 APIHelper 인터페이스와 APIRepository 구현
- Retrofit 콜백 처리를 위한 NetworkResponse 클래스와 NetworkResponseListener 추가
## 특이 사항
- ViewHolder는 Adapter의 InnerClass로 구현
- ViewHoler에 대한 Click 이벤트 인터페이스 BaseRecyclerViewAdapter에 추가
- LocationSearchPresenter 에서 Presenter 소멸에 따른 에러처리 및 Retrofit 응답 에러 처리 필요
resolved #29

nusurprise committed:
ADD : Retrofit API, NetworkResponse, NetworkResponseListener

1. Retrofit API 싱글톤으로 구현
2. Retrofit 콜백 클래스 NetworkResponse, 콜백 리스너 NetworkResponseListener를 구현
3. Kakao 맵 Rest API 호출부 뼈대 구현
4. LocationListAdapter, SearchPresenter 클래스 생성

nusurprise committed:
Merge branch 'develop' into feature/SS01-Retrofit

nusurprise committed:
PULL : feature/OL02

commit my local develop before pulling origin/develop

nusurprise committed:
Merge branch 'develop' of https://github.com/boostcampth/boostcamp3_A into develop

nusurprise committed:
Merge branch 'develop' into feature/SS01-Retrofit

nusurprise committed:
ADD : Retrofit call and RecyclerView

1. Retrofit 콜백 처리를 위한 NetworkResponse, NetworkResponseListener 추가
2. Retrofit 호출을 담당하는 APIHelper 인터페이스와 구현체 APIRepository추가
3. LocationSearch Activity, Presenter, Contract, LocationListAdapter 추가

nusurprise committed:
Merge branch 'develop' of https://github.com/boostcampth/boostcamp3_A into develop

nusurprise committed:
Merge branch 'develop' into feature/SS01-Retrofit

nusurprise committed:
MODIFIED : LocationSearchActivity 초기화 부분 수정, Retrofit API key 저장

1. LocationSearchActivity 초기화 부분 setbackarrow(), activitybinding부분 수정
2. Kakao Rest API key 'strings.xml'에 저장

nusurprise commented on app/src/main/java/com/aone/menurandomchoice/views/main/MainActivity.java:
```diff
@@ -29,15 +32,16 @@ public void onClick(View view) {
 
         switch (view.getId()) {
             case R.id.btn_start:
+                startActivity(new Intent(MainActivity.this, LocationSearchActivity.class));
```
테스트를 위한 코드였습니다, 수정해야합니다..

ch-Yoon commented on app/src/main/java/com/aone/menurandomchoice/repository/DataRepository.java:
```diff
@@ -13,12 +17,14 @@
     private static Repository repository;
     private KakaoLoginHelper kakaoLoginHelper;
     private ServerDataHelper serverDataHelper;
+    private APIHelper apiHelper;
```
APIRepository를 APIHelper로 Rapping해서 사용하셨네요! 
class 자체를 직접 선언하는것이 아닌 interface를 통해 접근하는 것은
객체간의 결합도를 낮춰줍니다. 좋아보입니다.

ch-Yoon commented on app/src/main/java/com/aone/menurandomchoice/repository/network/APIRepository.java:
```diff
@@ -0,0 +1,54 @@
+
+    private APIRepository() {
+            String BASE_URL = "http://ec2-54-180-49-154.ap-northeast-2.compute.amazonaws.com:8080/";
+
```
현 프로젝트에서는 BASE_URL이 바뀌는 구간이 존재합니다. 그리고 현재 코드는 BASE_URL을 고정으로 박아놓고, BASE_URL이 다를경우 APIInteface에서 URL을 덮어쓰는 방향으로 되어있는것으로 확인했습니다.

구글링을 해보시면, BASE_URL이 다를 경우에 대한 대처법이 존재하는데, 현재 방법과 대처방법이 어떤 차이가 있고, 어떤 장단점이 존재하는지 확인해보시면 좋을 것 같습니다.

nusurprise replied:
AWS API에 접근하는 메소드의 @GET 에는 BASE_URL 없이 endpoint만 명시하고
Kakao API에 접근하는 메소드의 @GET 에는 full url을 명시하겠습니다

nusurprise replied:
Fixed in 7497da03738c89320c1123907d36e54d9a0fe2c7

ch-Yoon commented on app/src/main/java/com/aone/menurandomchoice/repository/network/APIRepository.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public Call<AddressResponseBody> executeLocationSearch(NetworkResponseListener<AddressResponseBody> listener, String Qeury, String REST_API_KEY) {
+            Call<AddressResponseBody> call = apiInstance.getAddress(REST_API_KEY, Qeury);
```
저희끼리 정한 코드컨벤션에서는, 매개변수에 listener가 존재한다면 listener를 매개변수 가장 끝에 위치하기로 했었는데요~!! 확인해주시면 좋을 것 같습니다 :)

nusurprise replied:
코딩컨벤션 순서 반영하겠습니다

ch-Yoon commented on app/src/main/java/com/aone/menurandomchoice/views/base/adapter/BaseRecyclerViewAdapter.java:
```diff
@@ -89,21 +90,40 @@ public int getItemCount() {
+                }
+            }
+        });
+
```
onClickListener는 onBindViewHolder가 아닌, onCreateViewHolder에서 달아주는것이 좋아보입니다.
RecyclerView에서의 Adapter는, view를 재활용하는 것이 아닌 새롭게 뷰를 inflate해줘야 할때 onCreateViewHolder가 불리고, 기존에 생성한 viewHolder를 재활용할 수 있다면, 그 viewHolder를 재활용하면서 onBindViewHolder가 불려집니다. 즉, onBindViewHolder는 기존에 생성된 viewHolder가 매개변수로 전달된다는 것을 의미합니다. 

onCreateViewHolder에서 listener를 달아주게되면 최초 생성시에만 listener를 달아주면 되지만, onBindeViewHolder에서 달아주게 되면, view가 화면에 노출될때마다 달아주게 되는것입니다. 확인 부탁드립니다 :)

nusurprise replied:
createViewHolder에서 viewHolder의 position을 넘겨주기 위해 getAdapterPosition() 사용하여 반영하겠습니다

ch-Yoon commented on app/src/main/java/com/aone/menurandomchoice/views/base/adapter/BaseRecyclerViewAdapter.java:
```diff
@@ -89,21 +90,40 @@ public int getItemCount() {
-
     @NonNull
-    abstract VH createViewHolder(@NonNull LayoutInflater layoutInflater, @NonNull ViewGroup viewGroup, int i);
+    abstract public VH createViewHolder(@NonNull LayoutInflater layoutInflater, @NonNull ViewGroup viewGroup, int i);
```
자식 class에서 사용하기 위해 접근지정자를 설정하셨다면, protected로 설정하시는 것은 어떨까요?

nusurprise replied:
접근지정자 반영하겠습니다

ch-Yoon commented on app/src/main/java/com/aone/menurandomchoice/views/base/adapter/BaseRecyclerViewAdapter.java:
```diff
@@ -89,21 +90,40 @@ public int getItemCount() {
+    public void setOnItemClickListener(
+            OnItemClickListener onItemClickListener) {
+        this.onItemClickListener = onItemClickListener;
+    }
```
외부에서 접근하는 함수에서는 @NonNull 혹은 @Nullable 키워드를 붙여주시는게 좋아보입니다

nusurprise replied:
@nonnull 어노테이션 반영하겠습니다

ch-Yoon commented on app/src/main/java/com/aone/menurandomchoice/views/base/adapter/BaseRecyclerViewAdapter.java:
```diff
@@ -89,21 +90,40 @@ public int getItemCount() {
     @Override
-    public VH onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
+    public VH onCreateViewHolder(@NonNull ViewGroup viewGroup, int position) {
         LayoutInflater layoutInflater = LayoutInflater.from(viewGroup.getContext());
```
i보다는 position이 가독성 측면에서 더욱 좋아보이네요~! 수정한 것이 좋아보입니다

ch-Yoon commented on app/src/main/java/com/aone/menurandomchoice/views/locationsearch/LocationSearchPresenter.java:
```diff
@@ -0,0 +1,52 @@
+        implements LocationSearchContract.Presenter {
+
+    private Call<AddressResponseBody> call; // presenter 소멸시에 Retorit 중단을 위한 객체
+    private LocationSearchAdapter adapter;
```
call 중단하는 작업을 하기 위해 현재 presenter에서 call 객체를 갖고잇는데요~! 혹시 이 부분을 presenter에서 갖고 있는 것이 아니라, repository쪽에서 갖고 있을 수는 없을까요? 현재 Presenter에서 call 객체를 갖고 있다는 것은, 'Presenter가 이미 repository에서 retrofit을 사용한다는 것을 알고있다'가 되는것 같습니다.

nusurprise replied:
APIRepository로 call 객체를 옮기고 Presenter가 외부에서 접근할 수 있도록 method를 제공하겠습니다

ch-Yoon commented on app/src/main/java/com/aone/menurandomchoice/views/locationsearch/LocationSearchPresenter.java:
```diff
@@ -0,0 +1,52 @@
+            }
+        });
+        this.adapter = adapter;
+    }
```
현재 Adapter를 Presenter에서 직접 들고있습니다.
LocationSearchAdpater를 위한 AdapterView, AdapterModel interface를 정의하셔서
AdapterView는 View에서, AdapterModel는 Presenter에서만 들고있는것은 어떨까요?

말씀드린 방법으로 된다면, View에서 listener를 달고, 이벤트 발생 시 presenter에게 요청하는 구조로 진행 될 수 있을 것 같습니다. 객체간의 책임이 좀더 분명하게 자리잡을 것 같은데, 어떨까요?

nusurprise replied:
BaseRecyclerViewAdapter 내부에 선언 되어 있던 OnClickListener Interface를 밖으로 빼내고 BaseRecyclerViewAdapterView Interface를 선언하겠습니다

ch-Yoon commented on app/src/main/java/com/aone/menurandomchoice/views/locationsearch/LocationSearchPresenter.java:
```diff
@@ -0,0 +1,52 @@
+                    updateList(response.getDocuments());
+                }
+            }, Query, REST_API_KEY);
+    }
```
위에서 말씀드린대로, call객체를 presenter에서 들고있는것이 아닌 repository 내부로 넣을 수 있다면, presenter에서 직접 response를 받는것이 아닌, repository로부터 List를 전달받을 수 있게 될 것 같습니다~!

nusurprise replied:
여기서 call 객체는 콜백 처리를 위한 목적이 아닌 Retrofit 작업 중단을 위해 위치하고 있습니다,
콜백 처리는 Callback을 상속받은 NetworkResponse 객체에서 처리하고 있습니다. NetworkResponse객체가 response.body()를 Presenter에서 생성된 NetworkResponseListener에 전달함으로써 Presenter는 바로 요청 객체만 받아 볼 수 있습니다.

ch-Yoon commented on app/src/main/java/com/aone/menurandomchoice/views/locationsearch/LocationSearchContract.java:
```diff
@@ -0,0 +1,20 @@
+        void setAdapter(LocationSearchAdapter adapter);
+
+        void requestLocationSearch(@NonNull String Query, @NonNull String REST_API_KEY);
+    }
```
@NonNull 키워드를 붙여주신것이 좋아보입니다.
위에 setAdapter에서도 붙여주시면 어떨가요?

nusurprise replied:
@nonnull 반영하겠습니다

ch-Yoon commented on app/src/main/java/com/aone/menurandomchoice/repository/network/APIHelper.java:
```diff
@@ -0,0 +1,13 @@
+            @NonNull NetworkResponseListener<AddressResponseBody> networkResponseListener,
+            String Query,
+            String REST_API_KEY);
+}
```
String도 객체이므로, @NonNull 혹은 @Nullable을 붙여주시는게 좋아보입니다.

nusurprise replied:
어노테이션 반영하겠습니다

ch-Yoon reviewed (commented):
고생하셨습니다 :)

nusurprise committed:
ADD : MenuLocationResponseBody, BaseRecyclerViewAdapterView

1. MenuLocationResponseBody (AWS API 서버 호출 응답 객체)
2. adapterView와 adapterModel 역할  분리를 위해 BaseRecyclerViewAdapterView 추가
3. Retrofit '@GET' endpoint 변경

nusurprise committed:
Merge branch 'develop' of https://github.com/boostcampth/boostcamp3_A into develop

nusurprise committed:
MODIFIED: MainActivity

clear test code

nusurprise committed:
Apply .gitignore

nusurprise committed:
Merge branch 'develop' of https://github.com/boostcampth/boostcamp3_A into develop

nusurprise committed:
Merge branch 'develop' into feature/SS01-Retrofit

ParkSangGwon commented:
rebuild please

ParkSangGwon committed:
Merge branch 'develop' of https://github.com/boostcampth/boostcamp3_A into feature/SS01-Retrofit

nusurprise committed:
MODIFIED : resolve conflict

resolve conflict

ch-Yoon reviewed (approved):
고생하셨습니다
